### DISCUSSION TITLE
Shift all components to use React on a monorepo type setup

### DISCUSSION THREAD

**Discussion by spacemonkd** [apache/ozone#7108] [Ideas]
As we are having more changes in the UI and revamping the overall look of Recon, I wanted to take a look at the current frameworks that are being used in Ozone and if we can make it more efficient by using a single framework.

## Motivation
Looking at the current build times, and the overall package size I feel it could be optimized. Also from a code point of view, having  different frameworks in the same repository is hard to maintain and often things like package upgrades and fixes in dependencies go missing in some cases.
This increases the amount of work we might need to do later in case we are requiring to do any such task.

## Proposed Change
I want to propose to use a monorepo structure, where we use React as the UI framework for all the components (i.e SCM, OM and Recon) , dividing it into workspaces, and thus having only few common dependencies in the project.
This might be a one-time effort to migrate the code, but in the long term we will have a more efficient way to build and deploy the project.
As a part of this:
- First we will need to setup the repo structure to use a root package.json instead of individual ones at each component level.
  - This will be having all common dependencies that are required like React, ReactDOM, Vite, Typescript etc.
  - Component level package.json will be adding extra dependencies as required by the component.
- Migrate SCM and OM to React and improve the existing UI

## Advantages vs Disadvantages
Advantages:
- Single tooling system for all components mean we can optimize the build process
- No need to take care of multiple dependencies across several components
- Easier on-boarding for contributors
  - People [like to use React more than Angular](https://survey.stackoverflow.co/2023/#worked-with-vs-want-to-work-with-webframe-worked-want), thus increasing the number of contributors who want to work on things
  - Better support in terms of documentation, libraries, tooling
  
Disadvantages:
- Existing contributors for SCM and OM UI might have to learn/understand React to get onboarded.
- Effort required to setup the project build, and migrate components

**Comment by errose28** [comment]
The proposal makes sense to me. I'm no web developer though, so I can't speak to any technical challenges this approach might have. The component WebUIs were largely inherited from Hadoop from what I remember, so it would be good to bring them up to speed with the rest of the Ozone repo. Just to clarify, this would only affect the front end of the web UIs, correct? The Java backend would remain unchanged?

If we go with this approach, I would also suggest doing it on a feature branch.

**Reply by spacemonkd** [reply]
Yes Ethan, you are correct.
This would only be affecting the frontend, and a few changes on pom.xml files for how we are copying over the built/transpiled JS files in target

**Comment by dombizita** [comment]
Thanks for proposing this @devabhishekpal! I think the project would definitely benefit from this. Agree with @errose28, feature branch is a good idea for this change.
